File: Data Structures & Algorithms/two-integer-sum/submission-4.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        dic1 = {}

        for i, n in enumerate(nums):
            diff = target - n
            if diff in dic1:
                return [dic1[diff], i]
            dic1[n] = i
        
        return []

            



        

File: Data Structures & Algorithms/combination-target-sum/submission-4.java
class Solution {
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> result = new ArrayList<>();
        backtrack(result, nums, target, new ArrayList<>(), 0, 0);
        return result;
    }

    private void backtrack(List<List<Integer>> result, int[] nums, int target, List<Integer> currList, int start, int sum) {
        if (sum == target) {
            result.add(new ArrayList<>(currList));
            return;
        }

        if (sum > target) {
            return;
        }

        for (int i = start; i < nums.length; i++) {
            currList.add(nums[i]);
            backtrack(result, nums, target, currList, i, sum + nums[i]);
            currList.remove(currList.size() - 1);
        }
    }
}
